hail/src/main/scala/is/hail/cxx/Emit.scala

```diff
@@ -857,6 +856,56 @@ class Emitter(fb: FunctionBuilder, nSpecialArgs: Int, ctx: SparkFunctionContext)
              |})
              |""".stripMargin)
 
+      case ir.NDArrayMap(child, elemName, body) =>
+        val elemPType = child.pType.asInstanceOf[PNDArray].elementType
+        val cxxElemType = typeToCXXType(elemPType)
+        val newElemPType = body.pType
+        val newDataContainer = PArray(newElemPType)
+
+        val ndt = emit(child)
+        val nd = fb.variable("nd", "NDArray", ndt.v)
+
+        val elemRef = fb.variable("elemRef", cxxElemType)
+        val bodyt = outer.emit(body,
+          env.bind(elemName, EmitTriplet(elemPType, "", "false", elemRef.toString, resultRegion)))
+
+        val length = fb.variable("length", "long", s"n_elements($nd.shape)")
+        val sab = resultRegion.arrayBuilder(fb, newDataContainer)
+        val data = fb.variable("data", "const char *")
+
+        val childPretty = StringEscapeUtils.escapeString(ir.Pretty.short(child))
+        val bodyPretty = StringEscapeUtils.escapeString(ir.Pretty.short(body))
+        present(
+          s"""
+             |({
+             | ${ nd.define }
+             | ${ length.define }
+             | ${ elemRef.define }
+             | ${ data.define }
+             |
+             | ${ ndt.setup }
+             | if (${ ndt.m }) {
+             |   ${ fb.nativeError("NDArray cannot be missing. IR: %s".format(childPretty)) }
+             | }
+             |
+             | ${ sab.start(length.toString) }
+             | for (auto i = 0; i < $length; ++i) {
+             |   $elemRef = load_element<$cxxElemType>($nd.data + i * $nd.elem_size);
```

Comment:
Are there plans to support e.g. slicing/filtering along dimensions? If so, you'll probably want to only map the elements in the data array that exist in the new array, right?

In that case, you're constructing a new NDArray where you assume all the data array elements are in the NDArray (since that's how `make_ndarray` works), but populating it with potentially-nonexistent elements.

This may be hard to test right now, since you don't currently have any IRs that support changing the shape/stride like that, but it would be great to see the logic handle that case now so it doesn't become a case that trips us up later.

Reply:
We're definitely going to be doing that soon. I can work on changing this up to be more robust.